File: leetcode/1603.cpp
// Runtime: 91 ms, faster than 42.18% of C++ online submissions for Design Parking System.
// Memory Usage: 33.2 MB, less than 32.24% of C++ online submissions for Design Parking System.
class ParkingSystem {
 public:
  enum class CarType_ { kBig = 1, kMedium = 2, kSmall = 3 };
  map<CarType_, int> spots_left_ = {}; // tally of cars
  ParkingSystem(int big, int medium, int small)
      : spots_left_({pair(CarType_::kBig, big), pair(CarType_::kMedium, medium),
                     pair(CarType_::kSmall, small)}) {}

  bool addCar(int carType) {
    if (spots_left_[static_cast<CarType_>(carType)] == 0) {
      return false;
    } else {
      spots_left_[static_cast<CarType_>(carType)] -= 1;
      return true;
    }
  }
};

/**
 * Your ParkingSystem object will be instantiated and called as such:
 * ParkingSystem* obj = new ParkingSystem(big, medium, small);
 * bool param_1 = obj->addCar(carType);
 */

File: leetcode/1603.py
"""
Runtime: 140 ms, faster than 67.89% of Python3 online submissions for Design Parking System.
Memory Usage: 14.9 MB, less than 25.92% of Python3 online submissions for Design Parking System.
"""
from collections import deque


class ParkingSystem:
    def __init__(self, big: int, medium: int, small: int):
        self.lot = [0, 0, 0]
        self.cap = [big, medium, small]

    def addCar(self, carType: int) -> bool:
        i = carType - 1
        spots_left = self.cap[i] - self.lot[i]
        if spots_left > 0:
            self.lot[i] += 1
            return True
        else:
            return False


# Your ParkingSystem object will be instantiated and called as such:
# obj = ParkingSystem(big, medium, small)
# param_1 = obj.addCar(carType)
